### DISCUSSION TITLE
Nginx externally redirect when try_files

### DISCUSSION THREAD

**Discussion by UchihaYuki** [nginx/nginx#180] [Q&A]
I use the [official nginx docker image](https://hub.docker.com/_/nginx). Following is my `default.conf.template`.

```
server {
    listen       9004; 

    root /usr/share/nginx/html;

    index index.html;

    location ~* \.(eot|ttf|woff|woff2|svg)$ {
        add_header Access-Control-Allow-Origin *;
    }

    location / {
        try_files $uri $uri/ /index.html;
    }
}
```
I have a file under `https://example.com/projects/index.html`. When I access to `https://example.com/projects`, it redirect me to `http://example.com:9004/projects/`.

Note: My `nginx` is behind `traefik`, an another reverse proxy.

Why it doesn't try `$uri/index.html` as specified in my configuration, but sent a 301 redirect with `$uri/`?

**Comment by UchihaYuki** [comment]
Even if I access `nginx` directly by`http://192.168.31.185:9004/projects`. It will send a 301 redirect to `http://192.168.31.185:9004/projects/`. Shouldn't it send me back the `/projects/index.html` directly when I access to `http://192.168.31.185:9004/projects`?

**Comment by thresheek** [comment]
Hi @UchihaYuki !

Can you try setting 'port_in_redirect off;' in your configuration?

See  https://nginx.org/en/docs/http/ngx_http_core_module.html#port_in_redirect for more details.

**Reply by UchihaYuki** [reply]
> Hi @UchihaYuki !
> 
> Can you try setting 'port_in_redirect off;' in your configuration?
> 
> See https://nginx.org/en/docs/http/ngx_http_core_module.html#port_in_redirect for more details.

Hi, @thresheek. Is it possible to do an internal redirect? It still does an external redirect. I hope it can send the file under `/usr/share/nginx/html/projects/index.html` directly when access `https://example.com/projects`, rather than send an external redirect.

**Comment by nanaya** [comment]
the `$uri/` matched and thus processed accordingly with the redirect. remove it from the list if the behaviour isn't wanted

**Comment by lonfme** [comment]
'absolute_redirect off;'

**Comment by Raphasha27** [comment]
`try_files` checks files on disk, but when none match, it does an **internal redirect** — not an external HTTP redirect. To get an external redirect (301/302), use `return` or `rewrite`:

```nginx
# Internal redirect (URI stays same in browser)
location / {
    try_files $uri $uri/ /index.html;
}

# External redirect (browser URL changes)
location / {
    try_files $uri $uri/ @redirect;
}

location @redirect {
    return 301 https://other-site.example$uri;
}
```

If you need `try_files` to trigger an external redirect conditionally:

```nginx
location /old/ {
    try_files $uri @external;
}

location @external {
    return 302 /new/$uri;
}
```

The key difference:
- **Internal redirect**: `try_files` → last fallback → `rewrite ... last` or `proxy_pass` — URL stays
- **External redirect**: `return 301/302 <url>` — browser gets redirect, URL changes

For a more complex setup with redirect based on file existence:

```nginx
location /assets/ {
    # Check if file exists, if not redirect to CDN
    try_files $uri $uri/ @cdn;
}

location @cdn {
    return 302 https://cdn.example.com$uri;
}
```

If you want to proxy instead of redirect, use `rewrite ... break` with `proxy_pass` in the named location instead of `return`.
